plugin/core/windows.py

```diff
@@ -379,45 +384,56 @@ def activate_view(self, view: ViewLike) -> None:
+        file_path = view.file_name() or ""
+
+        def needed_configs(configs: 'List[ClientConfig]') -> 'List[ClientConfig]':
+            new_configs = []
+            for c in configs:
+                if c.name not in self._sessions:
+                    new_configs.append(c)
+                elif all(not s.handles_path(file_path) for s in self._sessions[c.name]):
+                    debug('path not in existing {} session: {}'.format(c.name, file_path))
+                    new_configs.append(c)
+            return new_configs
+
         # have all sessions for this document been started?
         with self._initialization_lock:
-            new_configs = filter(lambda c: c.name not in self._sessions,
-                                 self._configs.syntax_configs(view, include_disabled=True))
+            new_configs = needed_configs(self._configs.syntax_configs(view, include_disabled=True))
+            # filter(lambda c: c.name not in self._sessions,
+            #                  self._configs.syntax_configs(view, include_disabled=True))
 
             if any(new_configs):
                 # TODO: cannot observe project setting changes
                 # have to check project overrides every session request
                 self.update_configs()
 
-                startable_configs = filter(lambda c: c.name not in self._sessions,
-                                           self._configs.syntax_configs(view))
+                startable_configs = needed_configs(self._configs.syntax_configs(view))
 
                 for config in startable_configs:
-                    debug("window {} requests {} for {}".format(self._window.id(), config.name, view.file_name()))
-                    self._start_client(config)
 
-    def _start_client(self, config: ClientConfig) -> None:
-        workspace = self._ensure_workspace()
+                    debug("window {} requests {} for {}".format(self._window.id(), config.name, file_path))
+                    self._start_client(config, file_path)
 
-        if workspace is None:
-            debug('Cannot start without a project folder')
-            return
+    def _start_client(self, config: ClientConfig, file_path: str) -> None:
 
-        if not self._can_start_config(config.name):
+        if not self._can_start_config(config.name, file_path):
             debug('Already starting on this window:', config.name)
             return
 
         if not self._handlers.on_start(config.name, self._window):
             return
 
-        project_path = workspace.path
+        # TODO: this throws away previous ad-hoc folders.
+        # The previous fix was to only set an ad-hoc folder once per window
+        self._workspace_folders = get_workspace_folders(self._window, file_path)
+
         self._window.status_message("Starting " + config.name + "...")
-        debug("starting in", project_path)
+        debug("starting in", self._workspace_folders[0].path)
```

Comment:
You can use
```
debug("starting in", self._workspace_folders[0])
```
since `WorkspaceFolder` has a `__str__`

Although shouldn't it be

```
debug("starting in", self._workspace_folders)
```
? I guess it depends on wether the session supports multi folders or not?

Reply:
It's a bit misleading, but I took this message as stating the working directory of the server process. 

That probably also could use some improvement, if you have `/project/A` and `/project/B` folders, the server should perhaps have its working directory at `/project`?

Reply:
Given that we don't always start an executable, perhaps that logging should be moved and we could keep the logging after initialization that confirms which folders are part of the session.

Reply:
> That probably also could use some improvement, if you have /project/A and /project/B folders, the server should perhaps have its working directory at /project?

I thought of `sublime.cache_path()/LSP/servername`, but that broke some assumptions for clangd. Perhaps the cwd should stay at the first workspace folder for now so as to not break any assumptions. Leave that problem for another PR :)